common/setups/hybrid/gmm_system.py

```diff
@@ -646,210 +682,535 @@ def recognition(
+            costa_args = copy.deepcopy(self.hybrid_init_args.costa_args)
+            if self.crp[all_c].language_model_config is None:
+                costa_args["eval_lm"] = False
+            self.costa(all_c, prefix="costa/", **costa_args)
+            if costa_args["eval_lm"]:
+                self.jobs[all_c]["costa"].update_rqmt("run", {"mem": 8, "time": 24})
 
         for trn_c in self.train_corpora:
             self.store_allophones(trn_c)
 
         for eval_c in self.dev_corpora + self.test_corpora:
             self.create_stm_from_corpus(eval_c)
-            self.set_sclite_scorer(eval_c)
-
-        if "extract" in steps:
-            self.extract_features(
-                feat_args=self.hybrid_init_args.feature_extraction_args
-            )
-
-        # ---------- Monophone ----------
-        if "mono" in steps:
-            for trn_c in self.train_corpora:
-                self.monophone_training(
-                    "mono",
-                    trn_c,
-                    self.gmm_monophone_args.linear_alignment_args,
-                    **self.gmm_monophone_args.monophone_training_args,
-                )
-
-                for dev_c in self.dev_corpora:
-                    feature_scorer = (trn_c, "train_mono")
-                    self.recognition(
-                        f"mono-{trn_c}-{dev_c}",
-                        corpus=dev_c,
-                        feature_scorer=feature_scorer,
-                        **self.gmm_monophone_args.monophone_recognition_args,
-                    )
-
-                for tst_c in self.test_corpora:
-                    pass
-
-        # ---------- CaRT ----------
-        if "cart" in steps:
-            for c in self.train_corpora:
-                self.cart_and_lda(
-                    "mono",
-                    c,
-                    alignment="train_mono",
-                    **self.gmm_triphone_args.cart_lda_args,
+            if self.hybrid_init_args.scorer == "kaldi":
+                self.set_kaldi_scorer(
+                    corpus=eval_c,
+                    mapping={"[SILENCE]": ""},
                 )
-
-        # ---------- Triphone ----------
-        if "tri" in steps:
-            for c in self.train_corpora:
-                self.triphone_training(
-                    "tri",
-                    c,
-                    initial_alignment="train_mono",
-                    **self.gmm_triphone_args.triphone_training_args,
+            else:
+                self.set_sclite_scorer(
+                    corpus=eval_c,
+                    sort_files=False,
                 )
 
-            for c in self.dev_corpora:
-                # recog and optimize
-                pass
-
-            for c in self.test_corpora:
-                pass
+        for step_idx, (step_name, step_args) in enumerate(steps.get_step_iter()):
+            # ---------- Feature Extraction ----------
```

Comment:
I think every step definition should be a function, this way it is easier when working with class inheritance.

Reply:
Yes I think this looks really good now!